packages/lib-webpack/src/webpack/DynamicRemotePlugin.ts

```diff
@@ -0,0 +1,167 @@
+   */
+  pluginMetadata: string | PluginBuildMetadata;
+
+  /**
+   * List of extensions contributed by the plugin.
+   *
+   * The value is either a `minimatch` compatible JSON file glob pattern,
+   * or the parsed extensions array.
+   *
+   * Default value: `extensions.json`.
+   */
+  extensions: string | Extension[];
+
+  /**
+   * Modules shared between the host application and its plugins at runtime.
+   *
+   * Essential modules like `react` and `redux` will be added automatically.
+   *
+   * Default value: empty object.
+   */
+  sharedModules: WebpackSharedModules;
+}>;
+
+export class DynamicRemotePlugin implements webpack.WebpackPluginInstance {
+  private readonly pluginMetadata: PluginBuildMetadata;
+
+  private readonly extensions: Extension[];
+
+  private readonly sharedModules: WebpackSharedModules;
+
+  constructor(options: DynamicRemotePluginOptions = {}) {
+    const adaptedOptions: Required<DynamicRemotePluginOptions> = {
+      pluginMetadata: options.pluginMetadata ?? 'plugin.json',
+      extensions: options.extensions ?? 'extensions.json',
+      sharedModules: options.sharedModules ?? {},
+    };
+
+    // Load and validate plugin metadata
+    if (typeof adaptedOptions.pluginMetadata === 'string') {
+      this.pluginMetadata = loadPluginMetadata(adaptedOptions.pluginMetadata);
+    } else {
+      this.pluginMetadata = validatePluginMetadata(adaptedOptions.pluginMetadata);
+    }
+
+    // Load and validate extensions contributed by the plugin
+    if (typeof adaptedOptions.extensions === 'string') {
+      this.extensions = loadExtensions(adaptedOptions.extensions);
+    } else {
+      this.extensions = validateExtensions(adaptedOptions.extensions);
+    }
+
+    // Create plugin vs. host application shared module configuration
+    this.sharedModules = {
+      ...adaptedOptions.sharedModules,
+      ...getSharedModulesWithStrictSingletonConfig(['react', 'redux']),
+    };
+  }
+
+  apply(compiler: webpack.Compiler) {
+    const logger = compiler.getInfrastructureLogger(DynamicRemotePlugin.name);
+    const containerName = this.pluginMetadata.name;
+
+    if (!compiler.options.output.publicPath) {
+      logger.warn(
+        'output.publicPath is not defined, which may cause plugin assets to not load properly in the browser',
+      );
+    }
+
+    if (!compiler.options.output.uniqueName) {
+      logger.info(`output.uniqueName will be set to ${containerName}`);
+      compiler.options.output.uniqueName = containerName;
+    }
+
+    // Generate webpack federated module container assets
+    new webpack.container.ModuleFederationPlugin({
+      name: containerName,
+      library: {
+        type: 'jsonp',
+        name: remoteEntryCallback,
+      },
+      filename: remoteEntryScript,
+      exposes: _.mapValues(
+        this.pluginMetadata.exposedModules || {},
+        (moduleRequest, moduleName) => ({
+          import: moduleRequest,
+          name: `exposed-${moduleName}`,
```

Comment:
I would allow ignoring this `name`, webpack by default in production mode exposes these modules as hash numbers, that's much better for CDN to dump the cache if such file is stored and requested.

Reply:
Yeah, webpack's default chunk naming strategy for production is deterministic IDs :arrow_right: numbers which are unique to the given compilation.

Adding `exposed-` prefix doesn't add or subtract uniqueness of the chunk IDs, while it makes the file names more meaningful for humans.